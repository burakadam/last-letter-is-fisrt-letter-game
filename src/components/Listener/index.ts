export interface IVoiceRecognitionResult {
  distance: number;
  recognitionResult: string;
  textToRecognize: string;
}

export * from './Listener';
