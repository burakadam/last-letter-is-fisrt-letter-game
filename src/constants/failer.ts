export const FAILER_PERCENTAGE = 30;
