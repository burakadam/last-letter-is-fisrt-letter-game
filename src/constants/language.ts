export const DEFAULT_LANGUAGE = 'tr-TR';
