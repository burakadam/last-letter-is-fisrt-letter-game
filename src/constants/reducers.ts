export const REDUCERS = {
  GAME: 'game',
};
