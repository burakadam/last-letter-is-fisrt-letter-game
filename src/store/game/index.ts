export interface IErrors {
  listening: string | null;
  speaking: string | null;
}

export interface IGameState {
  gameStage: 'beggining' | 'game' | 'result';
  playerTurn: 'guest' | 'host' | null;
  isStarted: boolean;
  isSpeaking: boolean;
  isListening: boolean;
  hostAnswers: string[];
  activeHostAnswer: string;
  questAnswers: string[];
  activeGuestAnswer: string;
  answeredList: string[];
  resultText: string;
  errors: IErrors;
}

export * from './selectors';
export * from './slice';
export { default as gameReducer } from './slice';
