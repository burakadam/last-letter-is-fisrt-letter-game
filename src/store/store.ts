import {
  combineReducers,
  configureStore,
  PreloadedState,
} from '@reduxjs/toolkit';

import { REDUCERS } from '@/constants/reducers';
import { gameReducer } from './game';

const rootReducer = combineReducers({
  [REDUCERS.GAME]: gameReducer,
});

export const store = configureStore({
  reducer: rootReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
