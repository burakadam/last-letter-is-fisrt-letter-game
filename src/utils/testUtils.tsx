import { PreloadedState } from '@reduxjs/toolkit';
import { render, RenderOptions } from '@testing-library/react';
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';

import { AppStore, RootState, setupStore } from '@/store/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function Wrapper({ children }: { children: React.ReactElement }) {
  return <>{children}</>;
}

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
