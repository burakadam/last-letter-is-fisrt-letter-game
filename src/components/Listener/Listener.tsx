import { useCallback, useEffect } from 'react';

import { DEFAULT_LANGUAGE } from '@/constants/language';
import { DELAY_AFTER_GUEST_ANSWER } from '@/constants/time';
import { gameActions, gameSelectors } from '@/store/game';
import { useAppDispatch, useAppSelector } from '@/store/hooks';
import { _checkGuestAnswer } from '@/utils/checkGuestAnswer';
import { _getLastName } from '@/utils/getLastName';
import { Timer } from '../Timer';

const SpeechRecognition =
  window.SpeechRecognition || window.webkitSpeechRecognition;

const Listener = () => {
  const dispatch = useAppDispatch();

  const playerTurn = useAppSelector(gameSelectors.playerTurn);
  const isListening = useAppSelector(gameSelectors.isListening);
  const activeHostAnswer = useAppSelector(gameSelectors.activeHostAnswer);
  const activeGuestAnswer = useAppSelector(gameSelectors.activeGuestAnswer);
  const answeredList = useAppSelector(gameSelectors.answeredList);

  const handleTimeUp = useCallback(() => {
    dispatch(
      gameActions.setResultText({ text: 'Belirlen sürede cevap veremedin' })
    );
    dispatch(gameActions.resetToResult());
  }, [dispatch]);

  useEffect(() => {
    if (playerTurn === 'guest' && !isListening && activeHostAnswer !== null) {
      try {
        let recognition = new SpeechRecognition();

        recognition.lang = DEFAULT_LANGUAGE;
        recognition.continuous = false;
        recognition.interimResults = false;
        recognition.maxAlternatives = 1;

        recognition.start();

        dispatch(gameActions.startListening());

        recognition.onresult = (result: any) => {
          const answer = _getLastName(result?.results[0][0].transcript) ?? null;

          dispatch(gameActions.setGuestActiveAnswer({ answer }));

          const { success, text } = _checkGuestAnswer(
            answer,
            activeHostAnswer,
            answeredList
          );

          if (success) {
            dispatch(gameActions.updateAnsweredList({ answer }));
            dispatch(gameActions.updateGuestAnswers({ answer }));
            dispatch(gameActions.stopListening());

            setTimeout(() => {
              dispatch(gameActions.setPlayerTurn({ turn: 'host' }));
              dispatch(gameActions.setHostActiveAnswer({ answer: null }));
            }, DELAY_AFTER_GUEST_ANSWER);
          } else {
            dispatch(gameActions.stopListening());

            setTimeout(() => {
              dispatch(gameActions.setResultText({ text }));
              dispatch(gameActions.resetToResult());
            }, DELAY_AFTER_GUEST_ANSWER);
          }
        };
      } catch (error) {
        dispatch(gameActions.stopListening());
        dispatch(gameActions.errorListening({ error: String(error) }));
      }
    }
  }, [
    activeHostAnswer,
    dispatch,
    answeredList,
    handleTimeUp,
    isListening,
    playerTurn,
  ]);

  return (
    <div>
      <div className="text-center">
        <h1
          className={`mb-4 font-bold text-lg ${
            playerTurn === 'guest' ? 'text-red-600' : ''
          }`}
        >
          Senin sıran
        </h1>
        {playerTurn === 'guest' && (
          <p className="uppercase bg-yellow-400 p-2 text-white font-bold text-center mb-4">
            {activeGuestAnswer ? activeGuestAnswer : '...dinliyorum'}
          </p>
        )}
        {isListening && !activeGuestAnswer && <Timer onTimeUp={handleTimeUp} />}
      </div>
    </div>
  );
};

export { Listener };
