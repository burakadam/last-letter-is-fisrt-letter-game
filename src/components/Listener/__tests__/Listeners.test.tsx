import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';

import { REDUCERS } from '@/constants/reducers';
import { IGameState, initialState } from '@/store/game';
import { renderWithProviders } from '@/utils/testUtils';

import { Listener } from '../Listener';

const MockInitialState: IGameState = {
  ...initialState,
  gameStage: 'game',
  playerTurn: 'guest',
  isStarted: true,
  isListening: true,
  activeGuestAnswer: 'Burak',
};

describe('add listener that get names from host and validate it', () => {
  let container: HTMLElement;
  beforeEach(() => {
    // eslint-disable-next-line testing-library/no-render-in-setup
    ({ container } = renderWithProviders(<Listener />, {
      preloadedState: {
        [REDUCERS.GAME]: MockInitialState,
      },
    }));
  });

  it('render component correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('find the name comes from mock state', () => {
    expect(screen.getByText('Burak')).toBeInTheDocument();
  });
});
