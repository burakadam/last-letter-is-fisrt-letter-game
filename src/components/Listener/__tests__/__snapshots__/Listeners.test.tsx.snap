// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add listener that get names from host and validate it render component correctly 1`] = `
<div>
  <div>
    <div
      class="text-center"
    >
      <h1
        class="mb-4 font-bold text-lg text-red-600"
      >
        Senin sÄ±ran
      </h1>
      <p
        class="uppercase bg-yellow-400 p-2 text-white font-bold text-center mb-4"
      >
        Burak
      </p>
    </div>
  </div>
</div>
`;
